# **Recommender systems**

Types of Recommender systems
* Popularity based (not personnel and by using some formula)
* Content based (it is similarity based, like multiple similar matches)
* Collaborative based (based on user interests, past history)
* Hybrid Recommender system

## **Content based filtering**

#### **Movie Recommender System**

Data Gathering:

* tmdb_5000_credits.csv
* tmdb_5000_movies.csv

Datasets Download links:
* TMDB_5000_credits: [Download](https://drive.google.com/file/d/1WzSNlmC4bdhejo_YgCL4ob7bZufHfDjz/view?usp=drive_link)
* TMDB_5000_movies: [Download](https://drive.google.com/file/d/1f-dBMWZ_IWhGbvtxPWO3QHThS-Jh1ysM/view?usp=drive_link)

In [1]:
# importing necessary libraries

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
# data reading

movies_path = "../datasets/tmdb_5000_movies.csv"
credits_path = "../datasets/tmdb_5000_credits.csv"

movies = pd.read_csv(movies_path)
credits = pd.read_csv(credits_path)

In [10]:
# data understanding

print("shape of the moveis data:", movies.shape)
movies.head(1)

shape of the moveis data: (4803, 20)


budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title  vote_average  vote_count  
0  Enter the World of Pandora.  Avatar           7.2       11800

In [11]:
print(f"Credits data shape: {credits.shape}")
credits.head(1)

Credits data shape: (4803, 4)


movie_id   title                                               cast  \
0     19995  Avatar  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [13]:
credits.head(1)['crew'].values

array(['[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cam

In [16]:
movies = movies.merge(credits, on='title')

In [17]:
movies.shape

(4809, 23)

In [18]:
movies.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  ... runtime  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title vote_average vote_count movie_id  \
0  Enter the World of Pandora.  Avatar          7.2      11800    19995   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  

[1 rows x 23 columns]

* Need to create tags for performing 'content based recommendation'
* So, need to select the important features for tags creation

Important features to be resided:

genres, id, keywords, title, overview, cast, crew

In [20]:
movies = movies[['movie_id', 'title', 'overview', 'genres', 'keywords', 'cast', 'crew']]

In [21]:
movies.head(2)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

Now we will create another dataframe with attributes ('movie_id', 'title', 'tags')

In [22]:
# missing data 

movies.isnull().sum()

movie_id    0
title       0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [23]:
movies.dropna(inplace=True)

In [24]:
movies.isnull().sum()

movie_id    0
title       0
overview    0
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [26]:
# checking for duplicates
movies.duplicated().value_counts()

False    4806
Name: count, dtype: int64

No duplicates in the data

In [27]:
# processing genres column
movies.iloc[0].genres

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

In [32]:
from ast import literal_eval

def convert_genres(obj):
    l = []
    for i in literal_eval(obj):
        l.append(i['name'])
    return l

In [35]:
movies['genres'] = movies['genres'].apply(convert_genres)

In [36]:
movies.head(2)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [48]:
movies.head(1).keywords.values.tolist()

['[{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "space war"}, {"id": 3388, "name": "space colony"}, {"id": 3679, "name": "society"}, {"id": 3801, "name": "space travel"}, {"id": 9685, "name": "futuristic"}, {"id": 9840, "name": "romance"}, {"id": 9882, "name": "space"}, {"id": 9951, "name": "alien"}, {"id": 10148, "name": "tribe"}, {"id": 10158, "name": "alien planet"}, {"id": 10987, "name": "cgi"}, {"id": 11399, "name": "marine"}, {"id": 13065, "name": "soldier"}, {"id": 14643, "name": "battle"}, {"id": 14720, "name": "love affair"}, {"id": 165431, "name": "anti war"}, {"id": 193554, "name": "power relations"}, {"id": 206690, "name": "mind and soul"}, {"id": 209714, "name": "3d"}]']

In [53]:
movies['keywords'] = movies['keywords'].apply(convert_genres)

In [54]:
movies.head(2)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [56]:
movies.head(1).cast.values

array(['[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "ge

In [57]:
def convert_cast(obj):
    l =[]
    counter = 0
    for i in literal_eval(obj):
        if counter != 3:
            l.append(i['name'])
            counter += 1
        else:
            break
    return l

In [59]:
movies['cast'] = movies['cast'].apply(convert_cast)

In [60]:
def fetch_director(obj):
    l =[]
    for i in literal_eval(obj):
        if i['job'] == 'Director':
            l.append(i['name'])
    return l

In [62]:
movies['crew'] = movies['crew'].apply(fetch_director)

In [63]:
movies.head(2)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   

                                               cast              crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]  [Gore Verbinski]

In [66]:
movies['overview'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

In [68]:
movies['overview'] = movies['overview'].apply(lambda x: x.split())

In [69]:
movies.head(2)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   
1  [ocean, drug abuse, exotic island, east india ...   

                                               cast              crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   [James Cameron]  
1     [Johnny Depp, Orlando Bloom, Keira Knightley]  [Gore Verbinski]

In [71]:
movies['genres'] = movies['genres'].apply(lambda x: [i.replace(' ', '') for i in x])

In [72]:
movies['keywords'] = movies['keywords'].apply(lambda x: [i.replace(' ', '') for i in x])

In [73]:
movies['cast'] = movies['cast'].apply(lambda x: [i.replace(' ', '') for i in x])
movies['crew'] = movies['crew'].apply(lambda x: [i.replace(' ', '') for i in x])

In [74]:
movies.head(2)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   
1  [Captain, Barbossa,, long, believed, to, be, d...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   
1                  [Adventure, Fantasy, Action]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   
1  [ocean, drugabuse, exoticisland, eastindiatrad...   

                                            cast             crew  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]   [JamesCameron]  
1     [JohnnyDepp, OrlandoBloom, KeiraKnightley]  [GoreVerbinski]

In [75]:
movies['tags'] = movies['overview'] + movies['genres'] + movies['keywords'] + movies['cast'] + movies['crew']

In [85]:
new_df = movies[['movie_id', 'title', 'tags']].copy()

In [86]:
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [87]:
new_df['tags'] = new_df['tags'].apply(lambda x: ' '.join(x))

In [88]:
new_df.head(2)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   

                                                tags  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...

In [89]:
new_df['tags'][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. Action Adventure Fantasy ScienceFiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d SamWorthington ZoeSaldana SigourneyWeaver JamesCameron'

In [90]:
new_df['tags'] = new_df['tags'].apply(lambda x: x.lower())

In [91]:
new_df.head(2)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   

                                                tags  
0  in the 22nd century, a paraplegic marine is di...  
1  captain barbossa, long believed to be dead, ha...

In [104]:
import nltk
from nltk.stem.porter import PorterStemmer

ps = PorterStemmer()

In [105]:
def stem(text):
    y = []
    for i in text.split():
        y.append(ps.stem(i))
    return ' '.join(y)

In [107]:
new_df['tags'] = new_df['tags'].apply(stem)

In [109]:
# vectorization

from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer(
    max_features=5000,
    stop_words='english'
)

In [110]:
vectors = cv.fit_transform(new_df['tags']).toarray()

In [111]:
cv.get_feature_names_out()

array(['000', '007', '10', ..., 'zone', 'zoo', 'zooeydeschanel'],
      shape=(5000,), dtype=object)

In [115]:
# generating the cosine similarities of the vectors 

from sklearn.metrics.pairwise import cosine_similarity

similarity = cosine_similarity(vectors)

In [117]:
print("Shape of similarity matrix:", similarity.shape)

Shape of similarity matrix: (4806, 4806)


In [143]:
def recommend(movie):
    movie_index = new_df[new_df['title'] == movie].index[0]
    distances = similarity[movie_index]
    indexes_5 = np.argsort(distances)[::-1][1:6]
    movies = new_df.iloc[indexes_5]['title'].values.tolist()
    print(f"The top 5 most similar movies for '{movie}' are:\n")
    for idx, data in enumerate(movies):
        print(f"{idx+1}. {data}")

In [146]:
# checking the working of recomemndation

recommend("Avatar")

The top 5 most similar movies for 'Avatar' are:

1. Aliens vs Predator: Requiem
2. Aliens
3. Falcon Rising
4. Independence Day
5. Titan A.E.


In [148]:
recommend("Batman Begins")

The top 5 most similar movies for 'Batman Begins' are:

1. The Dark Knight
2. Batman
3. Batman
4. The Dark Knight Rises
5. 10th & Wolf


## **Collaborative based filtering**

### **Book Recommender System**

In [1]:
# importing the libraries

import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.metrics.pairwise import cosine_similarity

Datasets Downloading:
* Books: [Download](https://drive.google.com/file/d/1A_PRjojEJxPch0gAGMw3MGT0nviIkmKI/view?usp=drive_link)
* Users: [Download](https://drive.google.com/file/d/1ypLuzL3mD1PyCMQRdo1gcYqR6lCUzvdo/view?usp=drive_link)
* Ratings: [Download](https://drive.google.com/file/d/1iajeOFoC7f79ZrZFp_7K11oL1MyZJYVU/view?usp=drive_link)

In [3]:
# importing data

books = pd.read_csv('../datasets/books.csv', low_memory=False)
users = pd.read_csv('../datasets/users.csv')
ratings = pd.read_csv('../datasets/Ratings.csv')

In [4]:
books.head()

ISBN  ...                                        Image-URL-L
0  0195153448  ...  http://images.amazon.com/images/P/0195153448.0...
1  0002005018  ...  http://images.amazon.com/images/P/0002005018.0...
2  0060973129  ...  http://images.amazon.com/images/P/0060973129.0...
3  0374157065  ...  http://images.amazon.com/images/P/0374157065.0...
4  0393045218  ...  http://images.amazon.com/images/P/0393045218.0...

[5 rows x 8 columns]

In [5]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [6]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [7]:
# size of the data

print(f"Shape of Books: {books.shape}")
print(f"Shape of the users: {users.shape}")
print(f"Shape of ratings: {ratings.shape}")

Shape of Books: (271360, 8)
Shape of the users: (278858, 3)
Shape of ratings: (1149780, 3)


In [8]:
# Null values analysis

books.isna().sum()

ISBN                   0
Book-Title             0
Book-Author            2
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [9]:
users.isna().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

'Age' has more null count (but the thing is, we don't need age column in the recommedation analysis)

So, the null values of 'Age' column has no effect

In [10]:
ratings.isna().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

In [11]:
# checking for the duplicates 

books.duplicated().sum()

np.int64(0)

In [12]:
users.duplicated().sum()

np.int64(0)

In [13]:
ratings.duplicated().sum()

np.int64(0)

#### **Popularity based Recommender System**

Need to get,

**First 50 books, whose average is more and should be rated by minimum 250 users.**

These books directly loaded, when the page is loaded

In [14]:
ratings_with_names = ratings.merge(books, on='ISBN')

In [18]:
num_rating_df = ratings_with_names.groupby('Book-Title')['Book-Rating'].count().reset_index()
num_rating_df.columns = ['Book-Title', 'num_ratings']
print(f"Shape of result is: {num_rating_df.shape}")
num_rating_df.head()

Shape of result is: (241071, 2)


Book-Title  num_ratings
0   A Light in the Storm: The Civil War Diary of ...            4
1                              Always Have Popsicles            1
2               Apple Magic (The Collector's series)            1
3   Ask Lily (Young Women of Faith: Lily Series, ...            1
4   Beyond IBM: Leadership Marketing and Finance ...            1

In [20]:
avg_rating_df = ratings_with_names.groupby('Book-Title')['Book-Rating'].mean().reset_index()
avg_rating_df.columns = ['Book-Title', 'avg_rating']
print(f"Shape of data: {avg_rating_df.shape}")
avg_rating_df.head()

Shape of data: (241071, 2)


Book-Title  avg_rating
0   A Light in the Storm: The Civil War Diary of ...        2.25
1                              Always Have Popsicles        0.00
2               Apple Magic (The Collector's series)        0.00
3   Ask Lily (Young Women of Faith: Lily Series, ...        8.00
4   Beyond IBM: Leadership Marketing and Finance ...        0.00

In [21]:
popular_df = num_rating_df.merge(avg_rating_df, on='Book-Title')
print(f"Shape of result is: {popular_df.shape}")
popular_df.head()

Shape of result is: (241071, 3)


Book-Title  num_ratings  avg_rating
0   A Light in the Storm: The Civil War Diary of ...            4        2.25
1                              Always Have Popsicles            1        0.00
2               Apple Magic (The Collector's series)            1        0.00
3   Ask Lily (Young Women of Faith: Lily Series, ...            1        8.00
4   Beyond IBM: Leadership Marketing and Finance ...            1        0.00

In [23]:
result = popular_df[popular_df['num_ratings']>=250].sort_values('avg_rating', ascending=False)

In [25]:
# The First 50 books with rated more that 250 people and has highest avg rating
top_50_results = result.iloc[:51]
top_50_results

Book-Title  ...  avg_rating
80434   Harry Potter and the Prisoner of Azkaban (Book 3)  ...    5.852804
80422        Harry Potter and the Goblet of Fire (Book 4)  ...    5.824289
80441      Harry Potter and the Sorcerer's Stone (Book 1)  ...    5.737410
80426   Harry Potter and the Order of the Phoenix (Boo...  ...    5.501441
80414    Harry Potter and the Chamber of Secrets (Book 2)  ...    5.183453
191612  The Hobbit : The Enchanting Prelude to The Lor...  ...    5.007117
187377  The Fellowship of the Ring (The Lord of the Ri...  ...    4.948370
80445   Harry Potter and the Sorcerer's Stone (Harry P...  ...    4.895652
211384     The Two Towers (The Lord of the Rings, Part 2)  ...    4.880769
219741                              To Kill a Mockingbird  ...    4.700000
183573                                  The Da Vinci Code  ...    4.642539
187880                 The Five People You Meet in Heaven  ...    4.551163
180556                             The Catcher in the Rye  ...    4.545657
196326                          The Lovely Bones: A Novel  ...    4.468726
764                                                  1984  ...    4.454225
144165                           Prodigal Summer: A Novel  ...    4.450593
128670                                         Neverwhere  ...    4.449057
206502                            The Secret Life of Bees  ...    4.447028
168719  Stupid White Men ...and Other Sorry Excuses fo...  ...    4.356890
223135  Tuesdays with Morrie: An Old Man, a Young Man,...  ...    4.354970
204387                The Red Tent (Bestselling Backlist)  ...    4.334716
191589               The Hitchhiker's Guide to the Galaxy  ...    4.328358
129379   Nickel and Dimed: On (Not) Getting By in America  ...    4.289552
93381                                       Into the Wild  ...    4.273810
63867                                      Fahrenheit 451  ...    4.264059
74750                           Girl with a Pearl Earring  ...    4.218631
136145                                          Outlander  ...    4.173145
233370  Where the Heart Is (Oprah's Book Club (Paperba...  ...    4.105983
156102                     Seabiscuit: An American Legend  ...    4.098182
107962                                         Life of Pi  ...    4.088855
176845                                     The Bean Trees  ...    4.087404
2281    A Child Called \It\": One Child's Courage to S...  ...    4.086792
8434                                       ANGELA'S ASHES  ...    4.075269
76343                                         Good in Bed  ...    4.055102
64931   Fast Food Nation: The Dark Side of the All-Ame...  ...    4.037383
12700                                       American Gods  ...    4.006623
161645                                 Skipping Christmas  ...    4.006211
105777  Left Behind: A Novel of the Earth's Last Days ...  ...    4.003145
189551    The Golden Compass (His Dark Materials, Book 1)  ...    4.000000
181679                                   The Color Purple  ...    3.964968
160336                               Silence of the Lambs  ...    3.960938
8752                                          About a Boy  ...    3.900763
158138                  Seven Up (A Stephanie Plum Novel)  ...    3.888489
175097  The Alchemist: A Fable About Following Your Dream  ...    3.875940
80069   Hard Eight : A Stephanie Plum Novel (A Stephan...  ...    3.825279
170101                       Suzanne's Diary for Nicholas  ...    3.820569
111073                                  Lord of the Flies  ...    3.818533
5664                              A Prayer for Owen Meany  ...    3.796610
212070   The Vampire Lestat (Vampire Chronicles, Book II)  ...    3.777409
233851       White Oleander : A Novel (Oprah's Book Club)  ...    3.772472
234740  Wicked: The Life and Times of the Wicked Witch...  ...    3.766871

[51 rows x 3 columns]

#### **Collaborative Based Recommender System**

For accurate results

We consider the data, where each user should rate atleast  200 books

In [29]:
x = ratings_with_names.groupby('User-ID')['Book-Rating'].count()>200
user_result_indexes = x[x].index

In [30]:
filtered_rating_with_user = ratings_with_names[ratings_with_names['User-ID'].isin(user_result_indexes)]

In [33]:
y = filtered_rating_with_user.groupby('Book-Title')['Book-Rating'].count()>=50
filtered_books_indexes = y[y].index

In [34]:
final_ratings = filtered_rating_with_user[filtered_rating_with_user['Book-Title'].isin(filtered_books_indexes)]

In [37]:
pt = final_ratings.pivot_table(index='Book-Title', columns='User-ID', values='Book-Rating')

In [39]:
pt.fillna(0, inplace=True)

In [40]:
pt

User-ID                                             254     ...  278418
Book-Title                                                  ...        
1984                                                   9.0  ...     0.0
1st to Die: A Novel                                    0.0  ...     0.0
2nd Chance                                             0.0  ...     0.0
4 Blondes                                              0.0  ...     0.0
A Bend in the Road                                     0.0  ...     0.0
...                                                    ...  ...     ...
Year of Wonders                                        0.0  ...     0.0
You Belong To Me                                       0.0  ...     0.0
Zen and the Art of Motorcycle Maintenance: An I...     0.0  ...     0.0
Zoya                                                   0.0  ...     0.0
\O\" Is for Outlaw"                                    0.0  ...     0.0

[706 rows x 810 columns]

* considering each row as a vector with 810 dimentions
* We will apply the cosine similarity on each book with other books
* So at last we will get 706*706 matrix

In [41]:
similarity_score = cosine_similarity(pt)

In [42]:
similarity_score.shape

(706, 706)

In [68]:
def recommend(book_name):
    index = np.where(pt.index == book_name)[0][0]
    similar_books_indexes = np.argsort(similarity_score[index])[::-1][1:6]
    books = pt.index[similar_books_indexes].values
    print(f"The top {len(books)} books that are similar to {book_name} are:\n")
    for idx, book in enumerate(books):
        print(f"{idx+1}. {book}")

In [69]:
# performing collaborative recommendation

recommend('1984')

The top 5 books that are similar to 1984 are:

1. Animal Farm
2. The Handmaid's Tale
3. Brave New World
4. The Vampire Lestat (Vampire Chronicles, Book II)
5. The Hours : A Novel


In [70]:
recommend('Zoya')

The top 5 books that are similar to Zoya are:

1. Fine Things
2. Secrets
3. Kaleidoscope
4. Exclusive
5. Five Days in Paris


In [71]:
recommend('Harry Potter and the Chamber of Secrets (Book 2)')

The top 5 books that are similar to Harry Potter and the Chamber of Secrets (Book 2) are:

1. Harry Potter and the Prisoner of Azkaban (Book 3)
2. Harry Potter and the Goblet of Fire (Book 4)
3. Harry Potter and the Sorcerer's Stone (Harry Potter (Paperback))
4. Harry Potter and the Sorcerer's Stone (Book 1)
5. Harry Potter and the Order of the Phoenix (Book 5)


In [72]:
recommend('Life of Pi')

The top 5 books that are similar to Life of Pi are:

1. Tell No One
2. The Lovely Bones: A Novel
3. Fried Green Tomatoes at the Whistle Stop Cafe
4. Fall On Your Knees (Oprah #45)
5. Balzac and the Little Chinese Seamstress : A Novel
